DataFormats/Common/interface/getThinned_implementation.h

```diff
@@ -0,0 +1,172 @@
+#ifndef DataFormats_Common_getThinned_implementation_h
+#define DataFormats_Common_getThinned_implementation_h
+
+#include <optional>
+#include <tuple>
+
+#include "DataFormats/Common/interface/ThinnedAssociation.h"
+#include "DataFormats/Provenance/interface/BranchID.h"
+#include "DataFormats/Provenance/interface/ProductID.h"
+#include "DataFormats/Provenance/interface/ThinnedAssociationsHelper.h"
+
+namespace edm {
+  class WrapperBase;
+
+  namespace detail {
+    // This function provides a common implementation of
+    // EDProductGetter::getThinnedProduct() for EventPrincipal,
+    // DataGetterHelper, and BareRootProductGetter.
+    //
+    // getThinnedProduct assumes getIt was already called and failed to find
+    // the product. The input key is the index of the desired element in the
+    // container identified by ProductID (which cannot be found).
+    // If the return value is not null, then the desired element was
+    // found in a thinned container. If the desired element is not
+    // found, then an optional without a value is returned.
+    template <typename F1, typename F2, typename F3>
+    std::optional<std::tuple<WrapperBase const*, unsigned int> > getThinnedProduct(
+        ProductID const& pid,
+        unsigned int key,
+        ThinnedAssociationsHelper const& thinnedAssociationsHelper,
+        F1&& pidToBid,
+        F2&& getThinnedAssociation,
+        F3&& getByProductID) {
```

Comment:
On further thought I'm not sure if the perfect forwarding gives any real benefit here. The argument lambdas contain a pointer and possibly a 64-bit integer, so they don't benefit from moves, and copying them is cheap. Subsequent development introduces another recursive call within the body, which makes the use of `std::forward<>` more complex and error prone.